ReadMe.md
## bomWater

This `R` package grabs data from the Australian Bureau of Meteorology Water Data online (http://bom.gov.au/waterdata/) via the WISKI API.

Several functions have been written to retrieve quality checked timeseries. These are:

- `get_as_stored()`
- `get_hourly()`
- `get_daily()`
- `get_monthly()`
- `get_yearly()`

The following Water Data Online variables can be accessed using these functions:

| Parameter                      | Units  |
| ------------------------------ | ------ |
| Water Course Discharge         | m3/s   |
| Water Course Level             | m      |
| Electrical conductivity at 25C | µS/cm  |
| Turbidity                      | NTU    |
| pH                             | pH     |
| Water Temperature              | ºC     |
| Storage Volume                 | ML     |
| Storage Level                  | m      |
| Ground Water Level             | m      |
| Rainfall                       | mm     |
| Evaporation                    | mm     |
| Dry Air Temperature            | ºC     |
| Relative Humidity              | %      |
| Wind Speed                     | m/s    |

Make sure formatting of the parameter types is as in the table above when making requests. The function `bw_parameters()` can be used to retrieve this from within R.

Station information can be queried using `get_station_list()`, as in the example below.

The SOS2 manual can be consulted for the units of the different timeseries, as well as the meanings of the different quality codes ([BoM WISKI manual](http://www.bom.gov.au/waterdata/wiski-web-public/Guide\%20to\%20Sensor\%20Observation\%20Services\%20(SOS2)\%20for\%20Water\%20Data\%20\%20Online\%20v1.0.1.pdf)).

### Installation

```r
# Install the development version from GitHub
devtools::install_github('buzacott/bomWater')
```

### Copyright

All code is licensed MIT.

The license and copyright for the data can be viewed under the copyright tab at http://www.bom.gov.au/waterdata/. More information about the Bureau of Meteorology copyright can be found at http://www.bom.gov.au/other/copyright.shtml.

### Examples

```r
library(bomWater)

# Daily streamflow from Cotter River at Gingera (in m3/s)
cotter_river <- get_daily(parameter = 'Water Course Discharge',
                          station_number = '410730',
                          start_date     = '2020-01-01',
                          end_date       = '2020-01-31')

cotter_river
# A tibble: 31 x 3
   Timestamp           Value `Quality Code`
   <dttm>              <dbl>          <int>
 1 2020-01-01 00:00:00 0.013             10
 2 2020-01-02 00:00:00 0.013             10
 3 2020-01-03 00:00:00 0.011             10
 4 2020-01-04 00:00:00 0.009             10
 5 2020-01-05 00:00:00 0.01              10
 6 2020-01-06 00:00:00 0.015             10
 7 2020-01-07 00:00:00 0.023             10
 8 2020-01-08 00:00:00 0.019             10
 9 2020-01-09 00:00:00 0.017             10
10 2020-01-10 00:00:00 0.014             10
# … with 21 more rows

# Monthly total rainfall in mm at Cotter Hut
cotter_hut <- get_monthly(parameter = 'Rainfall',
                          station_number = '570946',
                          start_date     = '2019-01-01',
                          end_date       = '2019-12-31')
cotter_hut
# A tibble: 12 x 3
   Timestamp           Value `Quality Code`
   <dttm>              <dbl>          <int>
 1 2019-01-01 00:00:00  57.2             10
 2 2019-02-01 00:00:00  23.2             10
 3 2019-03-01 00:00:00  89.2             10
 4 2019-04-01 00:00:00  11.2             10
 5 2019-05-01 00:00:00 111.              10
 6 2019-06-01 00:00:00  44.8             10
 7 2019-07-01 00:00:00  38               10
 8 2019-08-01 00:00:00  50.8             10
 9 2019-09-01 00:00:00  50.8             10
10 2019-10-01 00:00:00  53.6             10
11 2019-11-01 00:00:00  41.2             10
12 2019-12-01 00:00:00   8               10

# Get a list of groundwater bore data available from water data online
get_station_list(parameter = 'Ground Water Level')
# A tibble: 4,439 x 5
   station_name station_no station_id station_latitude station_longitude
   <chr>        <chr>           <int>            <dbl>             <dbl>
 1 01/DD01 D    60930131       387998            -33.2              118.
 2 01/DD01 OB   60930132       388003            -33.2              118.
 3 01/DD04 S    60930135       388008            -33.2              118.
 4 02/DD25 OB   60930141       388013            -33.2              118.
 5 02/DD26 OB   60930142       388018            -33.2              118.
 6 02/DD27 OB   60930143       388023            -33.2              118.
 7 02/DD28 OB   60930144       388028            -33.2              118.
 8 02/DD29 OB   60930145       388033            -33.2              118.
 9 02/DD30 OB   60930146       388038            -33.2              118.
10 02/DD31 OB   60930147       388043            -33.2              118.
# … with 4,429 more rows
```


## Design philosphy
- Intentionally designed to simplify requests and output. For fine control use make_Bom_request to construct your own calls
- Most requests are determined by a combination of station_number and parameter. If multiple values are allowed for these fields, the argument is pluralised
   e.g. station_numbers, parameters
- All 'get' functions always return a tibble (including single column response)
   - if there's no result, a zero-row tibble is returned
   - the columns will correspond to the values (and order) provided by 'return_fields'
      - the exception is 'custom_attributes', which I need to figure out how to deal with
   - request fields can also be included in the return fields???????
   - if multiple stations are requested and not all stations have returned a result a warning is raised
- metadata functions return a tibble by default, can return a vector with specific arguments
   - these functions need a naming convention:
      - bw_parameters
      - bw_interpolation_types
      - bw_return_fields
      - bw_quality_codes
- helper functions which extract a specific column are named with the prefix extract_ and return a vector
   - Not user facing
   - extract_data_owner
   - extract_timezone
   - extract_timeseries_id
- results and calls are always in snake_case (syntactic)
   - requests may also use title case? actually no, this just adds complexity
   - interaction with parameters: e.g. station_number and parameter are uniform, and converted to bom names before request
   - in general, outputs are based on what you see on bom water data online
   - show how to convert to sentene case in vignette


NEWS.md
# bomWater 0.4.1

* New function to return a parameter list by station id added
* Improved docs for return_fields in all functions
* Examples are now not run due to possible loss of internet connectivity or Water Data Online going down for maintenance (which happens often in the evenings)
* Uses httptest in tests on CRAN with mock api calls due to avoid errors to due Water Data Online being down. tests also check for internet connectivity
* Uses httptest with vignette building for the same reason
* Fixed url in timeseries docs to fix a note during CRAN tests

# bomWater 0.4.0

* Converted style of functions to `tidyverse` style
* Added vignette
* Added tests for the core request functions

# bomWater 0.3.0

* Fixed an error that could occur from multiple tsIDs
* Fixed dependencies
* Fixed licence
* Fixed documentation and internal functions
* Added `parameters()` function to retrieve BoM parameters from `bomWater`

# bomWater 0.2.0

* First push to github
* Converted docs to roxygen
* Added var options for daily flow


R/getSeries.R
#' @title Get time series
#' @md
#' @description Get timeseries data from Water Data online
#' @details This function can be used if you want to retrieve a specific
#' timeseries that is not the default quality checked one.
#'
#' Common valid return fields are:
#'
#' * Timestamp
#' * Value
#' * Quality Code
#' * Interpolation Type
#'
#' Other valid return fields (depending on the parameter requested) may be:
#'
#' * Absolute Value
#' * AV Interpolation
#' * AV Quality Code
#' * Runoff Value
#' * RV Interpolation
#' * RV Quality Code
#' * Aggregation
#' * Accuracy
#'
#' If the request is not valid it will fail.
#' @param parameter_type The water data parameter type (e.g. Water Course
#' Discharge). See \code{\link{parameters()}} for a full list.
#' @param station_number The AWRC station number.
#' @param start_date Start date formatted as a string or date class
#' (YYYY-MM-DD).
#' @param end_date End date formatted as a string or date class (YYYY-MM-DD).
#' @param tz Optional: the desired time zone for the output timeseries. Input
#' must be an Olson Name (see `OlsonNames()`). By default the timeseries are
#' returned in non-DST time zones (AEST, ACST or AWST) depending on the
#' station location.
#' @param return_fields Optional: columns to be returned from Water Data Online.
#' By default Timestamp, Value and Quality Code are returned.
#' @param ts_name The timeseries name (e.g. DMQaQc.Merged.DailyMean.24HR) that
#' is desired.
#' @return
#' A tibble with columns with the requested return_fields. A zero row tibble is
#' returned if no data is returned  from the query. The columns of the tibble
#' are returned as character classes and have not been formatted to more
#' appropriate correct classes (this happens in other functions).
#' @seealso
#' * \url{http://www.bom.gov.au/waterdata/}
#' * \href{http://www.bom.gov.au/waterdata/wiski-web-public/Guide%20to%20Sensor%20Observation%20Services%20(SOS2)%20for%20Water%20Data%20%20Online%20v1.0.1.pdf}{BoM Guide to Sensor Observation Services (SOS2) for Water Data Online}
#' @examples
#' # Accessible dam storage, as shown on the BoM Water Storage dashboard
#' \dontrun{
#' get_timeseries(
#'   parameter_type = "Storage Volume",
#'   "G8150011",
#'   "2020-01-01",
#'   "2020-01-31",
#'   ts_name = "PR02AVQaQc.Merged.DailyMean.24HR",
#'   tz = NULL,
#'   return_fields = c("Timestamp", "Value", "Quality Code")
#' )
#' }
#' # See the linked SOS2 manual in See Also to find more timeseries names
#' @export

get_timeseries <- function(parameter_type,
                           station_number,
                           start_date,
                           end_date,
                           tz = "UTC",
                           return_fields,
                           ts_name) {

  # If no tz supplied, get from jurisdiction
  # All of this is currently skipped by the default field
  if (is.null(tz)) {
    # From BoM website:
    # Which time zones are the data displayed in?
    #   Time of day is presented in local standard time. Coordinated Universal Timezones (UTC) are:
    #
    #   Eastern States (QLD, NSW, ACT, VIC, TAS) - UTC +10:00
    #   Central States (NT, SA) - UTC +09:30
    #   Western Australia - UTC +08:00.

    # Get the station list and custom attributes to determine correct time zone
    station_list <- get_station_list(
      parameter_type = parameter_type,
      station_number = station_number,
      return_fields = "custom_attributes"
    )
    if (nrow(station_list) == 0) {
      stop(paste("Station number", station_number, "is invalid"))
    }

    jurisdiction <- stringr::str_split_fixed(station_list$DATA_OWNER_NAME, " -", n = 2)[1]

    # Time zones are selected where there is no DST
    if (jurisdiction %in% c("ACT", "ACTNSW", "NSW", "QLD", "TAS", "VIC")) {
      tz <- "Australia/Queensland" # AEST
    } else if (jurisdiction %in% c("SA", "NT")) {
      tz <- "Australia/Darwin" # ACST
    } else if (jurisdiction == "WA") {
      tz <- "Australia/Perth" # AWST
    } else {
      message("Jurisdiction not found, returning datetimes in UTC")
      tz <- "UTC"
    }
  } else {
    # Check if tz is valid
    if (!tz %in% OlsonNames()) {
      stop("Invalid tz argument. Check it is in OlsonNames().")
    }
    station_list <- get_station_list(parameter_type, station_number)
    if (nrow(station_list) == 0) {
      stop(paste("Station number", station_number, "is invalid"))
    }
  }

  # Check string date input is valid
  if (is.character(start_date)) {
    start_date <- lubridate::as_date(start_date, format = "%Y-%m-%d")
  }
  if (is.character(end_date)) {
    end_date <- lubridate::as_date(end_date, format = "%Y-%m-%d")
  }
  if (anyNA(c(start_date, end_date))) {
    stop("Dates must be formatted as %Y-%m-%d (e.g. 2000-01-01)")
  }
  # Ensure start is less than end
  if (start_date > end_date) {
    stop("start_date must be less than end_date")
  }

  # Coerce to datetime
  if (lubridate::is.Date(start_date)) {
    start_date <- lubridate::force_tz(lubridate::as_datetime(start_date), tz = tz)
    end_date <- lubridate::force_tz(lubridate::as_datetime(end_date), tz = tz)
  } else if (!lubridate::is.POSIXt(start_date)) {
    stop("Provide dates as a character, date, or datetime object")
  }

  # Fix the offset because BoM expects 00:00 style
  # despite 0000 conforming to ISO8601
  start_date <- sub(
    "(.*\\+)(\\d{2})(\\d{2})", "\\1\\2:\\3",
    lubridate::format_ISO8601(start_date, usetz = TRUE)
  )
  end_date <- sub(
    "(.*\\+)(\\d{2})(\\d{2})", "\\1\\2:\\3",
    lubridate::format_ISO8601(end_date, usetz = TRUE)
  )

  # Only accept one station at a time for now
  if (length(station_number) > 1) {
    stop("Only a single station can be requested at a time")
  }

  # If return_fields are missing return Timestamp, Value and Quality Code
  if (missing(return_fields)) {
    return_fields <- c("Timestamp", "Value", "Quality Code")
  }

  timeseries_id <- get_timeseries_id(
    parameter_type = parameter_type,
    station_number = station_number,
    ts_name = ts_name
  )

  timeseries_values <- get_timeseries_values(
    ts_id = timeseries_id$ts_id[1],
    start_date = start_date,
    end_date = end_date,
    return_fields = return_fields
  )

  # Only process data if it exists
  if (nrow(timeseries_values) > 0) {
    if ("Timestamp" %in% colnames(timeseries_values)) {
      # nolint start
      suppressMessages({
        timeseries_values$Timestamp <- lubridate::as_datetime(timeseries_values$Timestamp, tz = tz)
      })
      timeseries_values <- dplyr::mutate_at(timeseries_values,
        dplyr::vars(-"Timestamp"),
        utils::type.convert,
        as.is = TRUE
      )
      # nolint end
    } else {
      timeseries_values <- dplyr::mutate_all(timeseries_values,
        utils::type.convert,
        as.is = TRUE
      )
    }
  }
  return(timeseries_values)
}

#' @template timeseriesDocs
#' @examples
#' # Groundwater level as stored by the BoM
#' # PLUMB RD @ NARRABRI'
#' \dontrun{
#' get_as_stored(
#'   parameter_type = "Ground Water Level",
#'   station_number = "GW971623.3.3",
#'   start_date = "2020-03-01",
#'   end_date = "2020-03-01"
#' )
#' }
#' @export
get_as_stored <- function(parameter_type,
                          station_number,
                          start_date,
                          end_date,
                          tz = NULL,
                          return_fields = c("Timestamp", "Value", "Quality Code")) {
  parameter_type <- parameters()[tolower(parameter_type) == tolower(parameters())]
  if (length(parameter_type) == 0) {
    stop("Invalid parameter requested")
  }

  timeseries_values <- get_timeseries(
    parameter_type = parameter_type,
    station_number = station_number,
    start_date = start_date,
    end_date = end_date,
    tz = tz,
    return_fields = return_fields,
    ts_name = "DMQaQc.Merged.AsStored.1"
  )

  return(timeseries_values)
}

#' @template timeseriesDocs
#' @examples
#' # Hourly streamflow Cotter River at Gingera Gauge
#' \dontrun{
#' get_hourly(
#'   parameter_type = "Water Course Discharge",
#'   station_number = "410730",
#'   start_date = "2020-01-01",
#'   end_date = "2020-01-31"
#' )
#' }
#' @export
get_hourly <- function(parameter_type,
                       station_number,
                       start_date,
                       end_date,
                       tz = NULL,
                       return_fields = c("Timestamp", "Value", "Quality Code", "Interpolation Type")) {
  parameter_type <- parameters()[tolower(parameter_type) == tolower(parameters())]

  if (!parameter_type %in% c(
    "Water Course Discharge",
    "Water Course Level",
    "Storage Level",
    "Storage Volume"
  )) {
    stop(
      paste("Hourly data is not available for parameter_type", parameter_type)
    )
  }

  timeseries_values <- get_timeseries(
    parameter_type = parameter_type,
    station_number = station_number,
    start_date = start_date,
    end_date = end_date,
    tz = tz,
    return_fields = return_fields,
    ts_name = "DMQaQc.Merged.HourlyMean.HR"
  )

  return(timeseries_values)
}

#' @template timeseriesDocs
#' @param var The daily variable of interest. Valid inputs are `mean`, `min`,
#' `max` for continuous series such as discharge and `total` for discrete
#' series such as rainfall and evaporation.
#' @param aggregation Whether the data is to be aggregated midnight to
#' midnight (`24HR`) or from 9am-9am (`09HR`). The default is `24HR`. `09HR`
#' is only available for mean discharge and total rainfall and evaporation.
#' @examples
#' # Download daily mean aggregated over the standard day
#' \dontrun{
#' get_daily(
#'   parameter_type = "Water Course Discharge",
#'   station_number = "410730",
#'   start_date = "2020-01-01",
#'   end_date = "2020-01-31",
#'   var = "mean",
#'   aggregation = "24HR"
#' )
#' }
#'
#' # Download daily mean aggregated between 9am to 9am
#' \dontrun{
#' get_daily(
#'   parameter_type = "Water Course Discharge",
#'   station_number = "410730",
#'   start_date = "2020-01-01",
#'   end_date = "2020-01-31",
#'   var = "mean",
#'   aggregation = "09HR"
#' )
#' }
#'
#' # Download the daily max over the standard day
#' \dontrun{
#' get_daily(
#'   parameter_type = "Water Course Discharge",
#'   station_number = "410730",
#'   start_date = "2020-01-01",
#'   end_date = "2020-01-31",
#'   var = "max",
#'   aggregation = "24HR"
#' )
#' }
#'
#' @export
get_daily <- function(parameter_type,
                      station_number,
                      start_date,
                      end_date,
                      var = NULL,
                      aggregation = "24HR",
                      tz,
                      return_fields) {
  parameter_type <-
    parameters()[tolower(parameter_type) == tolower(parameters())]
  if (length(parameter_type) == 0) {
    stop("Invalid parameter requested")
  }

  # Handle possible formats of var input
  if (is.null(var)) {
    if (parameter_type %in% parameters("discrete")) {
      var <- "Total"
    } else {
      var <- "Mean"
    }
  } else {
    var <- stringr::str_to_title(var)
  }
  if (missing(aggregation)) {
    aggregation <- "24HR"
  } else {
    aggregation <- toupper(aggregation)
  }

  ts_name <- paste0("DMQaQc.Merged.Daily", var, ".", aggregation)

  if (parameter_type %in% parameters("continuous")) {
    valid_daily_ts <- c(
      "DMQaQc.Merged.DailyMean.24HR",
      "DMQaQc.Merged.DailyMax.24HR",
      "DMQaQc.Merged.DailyMin.24HR"
    )
    if (parameter_type == "Water Course Discharge") {
      valid_daily_ts <- c(
        valid_daily_ts,
        "DMQaQc.Merged.DailyMean.09HR"
      )
    }
  }

  if (parameter_type %in% parameters("discrete")) {
    valid_daily_ts <- c(
      "DMQaQc.Merged.DailyTotal.09HR",
      "DMQaQc.Merged.DailyTotal.24HR"
    )
  }

  if (!ts_name %in% valid_daily_ts) {
    stop("Invalid combination of parameter_type, var and aggregation")
  }

  if (missing(tz)) tz <- NULL

  if (missing(return_fields)) {
    return_fields <- c("Timestamp", "Value", "Quality Code")
  }

  timeseries_values <- get_timeseries(
    parameter_type,
    station_number,
    start_date,
    end_date,
    tz,
    return_fields,
    ts_name
  )

  return(timeseries_values)
}

#' @template timeseriesDocs
#' @examples
#' # Monthly average dry air temperature at Corin Dam
#' \dontrun{
#' get_monthly(
#'   parameter_type = "Dry Air Temperature",
#'   station_number = "570947",
#'   start_date = "2016-01-01",
#'   end_date = "2016-06-01"
#' )
#' }
#' @export
get_monthly <- function(parameter_type,
                        station_number,
                        start_date,
                        end_date,
                        tz,
                        return_fields) {
  parameter_type <-
    parameters()[tolower(parameter_type) == tolower(parameters())]
  if (length(parameter_type) == 0) {
    stop("Invalid parameter requested")
  }

  if (parameter_type %in% parameters("continuous")) {
    ts_name <- "DMQaQc.Merged.MonthlyMean.CalMonth"
  }

  if (parameter_type %in% parameters("discrete")) {
    ts_name <- c("DMQaQc.Merged.MonthlyTotal.CalMonth")
  }

  if (!exists("ts_name")) stop("Invalid parameter_type")

  if (missing(tz)) tz <- NULL

  if (missing(return_fields)) {
    return_fields <- c("Timestamp", "Value", "Quality Code")
  }

  timeseries_values <- get_timeseries(
    parameter_type,
    station_number,
    start_date,
    end_date,
    tz,
    return_fields,
    ts_name
  )

  return(timeseries_values)
}

#' @template timeseriesDocs
#' @param start_date Start date (formatted as YYYY-MM-DD) or just the
#' year (YYYY)
#' @param end_date End date (formatted as YYYY-MM-DD) or just the year (YYYY)
#' @examples
#' # Download annual rainfall for Cotter Hut
#' \dontrun{
#' get_yearly(
#'   parameter_type = "Rainfall",
#'   station_number = "570946",
#'   start_date = 2016,
#'   end_date = 2020
#' )
#' }
#'
#' @export
get_yearly <- function(parameter_type,
                       station_number,
                       start_date,
                       end_date,
                       tz,
                       return_fields) {
  parameter_type <-
    parameters()[tolower(parameter_type) == tolower(parameters())]
  if (length(parameter_type) == 0) {
    stop("Invalid parameter requested")
  }

  start_date <- paste0(stringr::str_sub(start_date, 1, 4), "-01-01")
  end_date <- paste0(stringr::str_sub(end_date, 1, 4), "-12-31")

  if (parameter_type %in% parameters("continuous")) {
    ts_name <- "DMQaQc.Merged.YearlyMean.CalYear"
  }

  if (parameter_type %in% parameters("discrete")) {
    ts_name <- c("DMQaQc.Merged.YearlyTotal.CalYear")
  }

  if (!exists("ts_name")) stop("Invalid parameter_type")

  if (missing(tz)) tz <- NULL

  if (missing(return_fields)) {
    return_fields <- c("Timestamp", "Value", "Quality Code")
  }

  timeseries_values <- get_timeseries(
    parameter_type,
    station_number,
    start_date,
    end_date,
    tz,
    return_fields,
    ts_name
  )

  return(timeseries_values)
}

#' @title Available water parameters
#' @description
#' `parameters` returns a vector of parameters that can be retrieved from
#' Water Data Online.
#' @param pars Optional: if empty all available parameters will be returned.
#' Alternatively, a vector of the continuous or discrete parameters can be
#' requested.
#' @return
#' A vector of parameters.
#' @details
#' The units of the parameters are as follows:
#'
#' * Water Course Discharge (m3/s)
#' * Water Course Level (m)
#' * Electrical conductivity at 25C (µS/cm)
#' * Turbidity (NTU)
#' * pH
#' * Water Temperature (ºC)
#' * Storage Volume (ML)
#' * Storage Level (m)
#' * Ground Water Level (m)
#' * Rainfall (mm)
#' * Evaporation (mm)
#' * Dry Air Temperature (ºC)
#' * Relative Humidity (%)
#' * Wind Speed (m/s)
#' @md
#' @seealso
#' * \url{http://www.bom.gov.au/waterdata/}
#' * \href{http://www.bom.gov.au/waterdata/wiski-web-public/Guide%20to%20Sensor%20Observation%20Services%20(SOS2)%20for%20Water%20Data%20%20Online%20v1.0.1.pdf}{BoM Guide to Sensor Observation Services (SOS2) for Water Data Online}
#' @examples
#' parameters()
#' parameters("continuous")
#' parameters("discrete")
#' @export
parameters <- function(pars) {
  continuous <- c(
    "Dry Air Temperature",
    "Relative Humidity",
    "Wind Speed",
    "Electrical Conductivity At 25C",
    "Turbidity",
    "pH",
    "Water Temperature",
    "Ground Water Level",
    "Water Course Level",
    "Water Course Discharge",
    "Storage Level",
    "Storage Volume"
  )
  discrete <- c(
    "Rainfall",
    "Evaporation"
  )

  if (missing(pars)) {
    return(c(discrete, continuous))
  } else {
    if (!tolower(pars) %in% c("continuous", "discrete")) {
      stop("Invalid parameter category entered")
    }
    return(get(pars))
  }
}
